lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,462 @@
+        [
+            line.strip("\n").strip().split()
+            for line in StringIO(
+                _dnf("list", postargs=[f"--{cache_type}"]).stdout.strip("\n"),
+            )
+        ],
+    ):
+        placeholder = pkg_name.split(".")
+        epoch = None  # Not every DNF package has an epoch.
+        charset, buffer = deque(itertools.chain(pkg_version)), []
+        while charset:
+            char = charset.popleft()
+            if char == ":":
+                epoch = "".join(buffer)
+                buffer.clear()
+            elif char == "-":
+                version = "".join(buffer)
+                buffer.clear()
+            else:
+                buffer.append(char)
+
+        result.append(
+            {
+                "name": ".".join(placeholder[:-1]),
+                "arch": ".".join(placeholder[-1:]),
+                "epoch": epoch,
+                "version": version,
+                "release": "".join(buffer),
+                "repo": pkg_repo[1:] if pkg_repo.startswith("@") else pkg_repo,
+                "state": state[cache_type],
+            }
+        )
+
+    return result
+
+
+class _Cache:
+    """Track the current state of the DNF package cache."""
+
+    def __new__(cls) -> "_Cache":
+        """Create new  _Cache object.
+
+        Returns:
+            (_Cache): New _Cache object.
+        """
+        if not hasattr(cls, f"{cls.__name__}__instance"):
+            cls.__instance = super(_Cache, cls).__new__(cls)
+            cls.__instance.__initialized = False
+        return cls.__instance
+
+    def __init__(self) -> None:
+        if self.__initialized:
+            return
+
+        self.__initialized = True
+        self.refresh()
+
+    @property
+    def installed(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get installed packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of installed packages.
+        """
+        return copy.deepcopy(self.__installed)
+
+    @property
+    def available(self) -> List[Dict[str, Union[str, _PackageState]]]:
+        """Get available packages.
+
+        Returns:
+            (List[Dict[str, Union[str, _PackageState]]]):
+                Cache of available packages.
+        """
+        return copy.deepcopy(self.__available)
+
+    def refresh(self) -> None:
+        """Refresh current package cache."""
+        with ThreadPoolExecutor() as executor:
+            installed = executor.submit(_load_cache, "installed")
+            available = executor.submit(_load_cache, "available")
+            self.__installed = installed.result()
+            self.__available = available.result()
+
+
+def _refresh(func: Callable) -> Callable:
+    """Refresh cache after function call.
+
+    Args:
+        func (Callable): Function to wrap.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs) -> None:
+        func(*args, **kwargs)
+        _Cache().refresh()
```

Comment:
I used the decorator so if I ever needed to change the refresh implementation, all I would have to do is modify the decorator. I can look into changing this however since we might want more granular controller over how the cache is refreshed.

Reply:
The decorator was removed in https://github.com/canonical/operator-libs-linux/pull/75/commits/661114b536c049933697507536f6270a1d2a6beb